metrics-core/src/main/java/com/codahale/metrics/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -73,24 +75,33 @@
             this.startTick = startTick;
             this.values = values;
             this.count = count;
+            this.priorityFloor = count >= size ? values.firstKey() : Double.NEGATIVE_INFINITY;
         }
 
         private void update(long value, long timestampNanos) {
             double itemWeight = weight(timestampNanos - startTick);
             double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
-            boolean mapIsFull = count >= size;
-            if (!mapIsFull || values.firstKey() < priority) {
-                addSample(priority, value, itemWeight, mapIsFull);
+            if (priorityFloor < priority) {
+                boolean mapIsFull = count >= size;
+                if (!mapIsFull || values.firstKey() < priority) {
```

Comment:
Might get a better hit rate on the fast path by updating `priorityFloor` here instead of upon removal.
```suggestion
                if (!mapIsFull || (priorityFloor = values.firstKey()) < priority) {
```